raiden/network/pathfinding.py

```diff
@@ -1,12 +1,146 @@
+
+    iou['signature'] = sign_one_to_n_iou(
+        privatekey=privkey,
+        expiration=iou['expiration_block'],
+        sender=iou['sender'],
+        receiver=iou['receiver'],
+        amount=iou['amount'],
+    )
+
+    return iou
+
+
+def create_current_iou(
+        config: typing.Dict[str, typing.Any],
+        token_network_address: typing.Union[typing.TokenNetworkAddress, typing.TokenNetworkID],
+        our_address: typing.Address,
+        privkey: str,
+        block_number: typing.BlockNumber,
+):
+    url = config['pathfinding_service_address']
+    latest_iou = get_pfs_iou(url, token_network_address)
+    if latest_iou is None:
+        return make_iou(config, our_address, privkey, block_number)
+    else:
+        added_amount = config['pathfinding_max_fee']
+        return update_iou(latest_iou, privkey, added_amount=added_amount)
+
+
+def query_paths(
+        service_config: typing.Dict[str, typing.Any],
+        our_address: typing.Address,
+        privkey: bytes,
+        current_block_number: typing.BlockNumber,
+        token_network_address: typing.Union[typing.TokenNetworkAddress, typing.TokenNetworkID],
+        route_from: typing.InitiatorAddress,
+        route_to: typing.TargetAddress,
+        value: typing.TokenAmount,
+):
+    max_paths = service_config['pathfinding_max_paths']
+    url = service_config['pathfinding_service_address']
+    payload = {'from': route_from, 'to': route_to, 'value': value, 'max_paths': max_paths}
+
+    payload.update(create_current_iou(
+        config=service_config,
+        token_network_address=token_network_address,
+        our_address=our_address,
+        privkey=privkey,
+        block_number=current_block_number,
+    ))
+
+    try:
+        response = requests.post(
+            f'{url}/api/v1/{token_network_address}/paths',
+            data=payload,
+            timeout=DEFAULT_HTTP_REQUEST_TIMEOUT,
+        )
+    except requests.RequestException:
+        raise ServiceRequestFailed(
+            'Could not connect to Pathfinding Service',
+            dict(parameters=payload, exc_info=True),
+        )
+
+    if response.status_code != 200:
+        info = {'error_code': response.status_code}
+        try:
+            error = response.json().get('errors')
+            if error is not None:
+                info['pfs_error'] = error
+        except ValueError:  # invalid json
+            pass
+        raise ServiceRequestFailed('Pathfinding service returned error code', info)
+
+    try:
+        return response.json()['result']
+    except KeyError:
+        raise ServiceRequestFailed(
+            "Answer from pathfinding service not understood ('result' field missing)",
+            dict(response=response.json()),
+        )
+    except ValueError:
+        raise ServiceRequestFailed(
+            'Pathfinding service returned invalid json',
+            dict(response_text=response.text, exc_info=True),
+        )
```

Comment:
Adding my comment here for the lack of a better place. This function does not seem to handle a case where the PFS returns an error code saying "I cashed out the previous IOU, create a new one and start from scratch"... did you get any documentation regarding this behaviour from the PFS-specs?

Reply:
No, currently all failed requests are handled by falling back to internal routing. I talked to Karl, and the error codes of the PFS are not yet defined and will be next week. So I want to merge this as is now and make a follow-up pr then.